src/detours.cpp

```diff
@@ -351,67 +361,77 @@ struct _DETOUR_TRAMPOLINE
     PF_DETOUR_INVOKE_CONTEXT_CALLBACK pfContextCallback;    // context callback used as indirect jmp in rbContext
+#endif // DETOURS_SUPPORT_DETOUR_CONTEXT_CALLBACK
 };
 
-C_ASSERT(sizeof(_DETOUR_TRAMPOLINE) == 168);
+#ifdef DETOURS_SUPPORT_DETOUR_CONTEXT_CALLBACK
+C_ASSERT(sizeof(_DETOUR_TRAMPOLINE) == 168);
+#else
+C_ASSERT(sizeof(_DETOUR_TRAMPOLINE) == 96);
+#endif // DETOURS_SUPPORT_DETOUR_CONTEXT_CALLBACK
 
 enum {
     SIZE_OF_JMP = 5
-};
-
-inline PBYTE detour_gen_context_callback(PBYTE pbCode, PF_DETOUR_INVOKE_CONTEXT_CALLBACK* ppfContextCallback, PVOID pContextParam)
-{
-    PBYTE pbJmpSrc = pbCode + 37;
-
-    // Backup volatile registers that are used to pass __fastcall params.
-    // If the set context function performed floating point operations, 
-    // registers XMM0, XMM1, XMM2, and XMM3 would also have to be backed 
-    // up and restored. In the attempt to keep this code generation as simple 
-    // as possible, floating point operations in the set context function isn't 
-    // currently supported.
-    *pbCode++ = 0x51;   // push rcx (backup rcx)
-    *pbCode++ = 0x52;   // push rdx (backup rdx)
-    *pbCode++ = 0x41;   // push r8 (backup r8)
-    *pbCode++ = 0x50;
-    *pbCode++ = 0x41;   // push r9 (backup r9)
-    *pbCode++ = 0x51;
-
-    *pbCode++ = 0x48;   // sub rsp,20h (required by x64 __fastcall)
-    *pbCode++ = 0x83;
-    *pbCode++ = 0xEC;
-    *pbCode++ = 0x20;
-
-    *pbCode++ = 0x48;   // mov rcx, imm64 (return address)
-    *pbCode++ = 0xB9;
-    *((INT64*&)pbCode)++ = (INT64)(pbJmpSrc);
-    *pbCode++ = 0x51;   // push rcx (return address)
-
-    *pbCode++ = 0x48;   // mov rcx,imm64 (param)
-    *pbCode++ = 0xB9;
-    *((INT64*&)pbCode)++ = (INT64)(pContextParam);
-
-    *pbCode++ = 0xFF;   // jmp [+imm32] (pfContextCallback)
-    *pbCode++ = 0x25;
-    *((INT32*&)pbCode)++ = (INT32)((PBYTE)ppfContextCallback - pbJmpSrc);
-
-    *pbCode++ = 0x48;   // add rsp,20h (required by x64 __fastcall)
-    *pbCode++ = 0x83;
-    *pbCode++ = 0xC4;
-    *pbCode++ = 0x20;
-
-    *pbCode++ = 0x41;   // pop r9 (restore r9)
-    *pbCode++ = 0x59;
-    *pbCode++ = 0x41;   // pop r8 (restore r8)
-    *pbCode++ = 0x58;
-    *pbCode++ = 0x5A;   // pop rdx (restore rdx)
-    *pbCode++ = 0x59;   // pop rcx (restore rcx)
-
-    return pbCode;
-}
+};
+
+#ifdef DETOURS_SUPPORT_DETOUR_CONTEXT_CALLBACK
+
+inline PBYTE detour_gen_context_callback(PBYTE pbCode, PF_DETOUR_INVOKE_CONTEXT_CALLBACK* ppfContextCallback, PVOID pContextParam)
+{
+    PBYTE pbJmpSrc = pbCode + 37;
+
+    // Backup volatile registers that are used to pass __fastcall params.
+    // If the set context function performed floating point operations, 
+    // registers XMM0, XMM1, XMM2, and XMM3 would also have to be backed 
+    // up and restored. In the attempt to keep this code generation as simple 
+    // as possible, floating point operations in the set context function isn't 
```

Comment:
Are you sure the compiler won't "unpredictably" use xmm0-3?
For example as part of an inlined constant size memset?

Reply:
Also we should try to create pdata/xdata for this stuff. Stepping every single instruction should be able to staclkwalk. This is how the system works w/o Detours. I don't want to turn down a PR w/o that..since you are presumably already shipping this.